### ISSUE TITLE
VSTHRD010 should not fire in methods that implement IVs* interface members

### ISSUE THREAD

**Issue by AArnott** [Microsoft/vs-threading#53]
It's a given that an IVs*Events interface will (almost certainly) be invoked on the UI thread. So requiring explicitly checking is wasteful, as shown here:

```csharp
// IVsSelectionEvents
public int OnCmdUIContextChanged(uint dwCmdUICookie, int fActive)
{
   ThreadHelper.ThrowIfNotOnUIThread(); // VSSDK002 should not demand this before other IVs* calls.
   ivssolution.DoSomething();
}
```

Similar argument for overrides of `Package.Initialize` or `Package.Dispose`.

**Comment by sharwell** [comment]
Yes, *provided* you expand VSSDK002 to demand that the check appear before calling methods that implement `IVs*`, even if that implementation isn't in a shell interop assembly.

**Comment by AArnott** [comment]
@sharwell: we already do that if the call is through the interface. But if the call isn't through the interface VSSDK002 won't catch it. Is that what you were referring to? Sounds like a good idea.

**Comment by AArnott** [comment]
(@sharwell: it's good to have you watching the project. I'm glad we OSS'd it if nothing than just for that)

**Comment by sharwell** [comment]
> But if the call isn't through the interface VSSDK002 won't catch it. Is that what you were referring to?

@AArnott Yes that's what I'm referring to. By removing the check in the implementation (this issue), you are assuming the caller is doing the right thing. But when it comes to calling the implementation, you aren't currently checking that. :smile:

**Comment by AArnott** [comment]
[This library already works in VS2012](https://blogs.msdn.microsoft.com/andrewarnottms/2014/05/07/asynchronous-and-multithreaded-programming-within-vs-using-the-joinabletaskfactory/#comments). We didn't need to OSS it for that. 

> The Microsoft.VisualStudio.Threading library is now available for Visual Studio 2010 and 2012, by installing it with your VS extension as described and allowed by this NuGet package: https://www.nuget.org/packages/Microsoft.VisualStudio.Threading.DownlevelInstaller

**Comment by AArnott** [comment]
I'm backtracking on this. I have made the change already for #65 that @sharwell suggested, that we detect calls to public methods that happen to implement `IVs*` interfaces and ensure that the caller asserts the UI thread first. But this issue's primary concern of *not* requiring a UI thread assert *inside* that method is what I've changed my mind about, because folks can call those methods directly (who may or may not have the analyzer) and may be on the wrong thread. So asserting the right thread helps protect against deadlocks.
